### PR TITLE
MLP: Add minibatch gradient descent solver

### PR DESCRIPTION
**Pull request by njayaram2**
JIRA: MADLIB-1206

This commit adds support for mini-batch based gradient descent for MLP.
If the input table contains a 2D matrix for independent variable,
minibatch is automatically used as the solver. Two minibatch specific
optimizers are also introduced: batch_size and n_epochs.
- batch_size is defaulted to min(200, buffer_size), where buffer_size is
  equal to the number of original input rows packed into a single row in
  the matrix.
- n_epochs is the number of times all the batches in a buffer are
  iterated over (default 1).

Other changes include:
- dependent variable in the minibatch solver is also a matrix now. It
  was initially a vector.
- Randomize the order of processing a batch within an epoch.
- MLP minibatch currently doesn't support weights param, an error is
  thrown now.
- Delete an unused type named mlp_step_result.
- Add unit tests for newly added functions in python file.

Co-authored-by: Rahul Iyer <riyer@apache.org>
Co-authored-by: Nikhil Kak <nkak@pivotal.io>

Closes #243

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/376/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/377/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/378/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/380/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/383/

File: src/ports/postgres/modules/convex/mlp_igd.py_in [hunk 1]
```diff
@@ -222,67 +243,83 @@ def mlp(schema_madlib, source_table, output_table, independent_varname,
...
                 # Initalize according to Glorot and Bengio (2010)
                 # See design doc for more info
                 span = math.sqrt(6.0 / (fan_in + fan_out))
-                dim = (layer_sizes[i] + 1) * layer_sizes[i + 1]
-                rand = plpy.execute("""SELECT array_agg({span}*2*(random()-0.5))
-                                       AS random
-                                       FROM generate_series(0,{dim})
-                        """.format(span=span, dim=dim))[0]["random"]
+                dim = (fan_in + 1) * fan_out
+                rand = [span * (random() - 0.5) for _ in range(dim)]
```
Comment (kaknikhil) [inline src/ports/postgres/modules/convex/mlp_igd.py_in]: why are we subtracting 0.5 from `random()` ?

File: src/ports/postgres/modules/convex/mlp_igd.py_in [hunk 2]
```diff
@@ -292,26 +329,33 @@ def mlp(schema_madlib, source_table, output_table, independent_varname,
...
         _update_temp_model_table(it_args, it.iteration, temp_output_table,
-                                 first_try)
+                                 is_minibatch_enabled, first_try)
         first_try = False
-    layer_sizes_str = py_list_to_sql_string(
-        layer_sizes, array_type="integer")
-    classes_str = py_list_to_sql_string(
-        [strip_end_quotes(cl, "'") for cl in classes],
-        array_type=dependent_type)
+    layer_sizes_str = py_list_to_sql_string(layer_sizes,
```
Comment (kaknikhil) [inline src/ports/postgres/modules/convex/mlp_igd.py_in]: should we use the `PY2SQL` alias here ?

File: src/ports/postgres/modules/convex/mlp_igd.py_in [hunk 3]
```diff
@@ -292,26 +329,33 @@ def mlp(schema_madlib, source_table, output_table, independent_varname,
...
-                                  warm_start)
+    if is_minibatch_enabled:
+        # We already have the mean and std in the input standardization table
+        input_std_table = add_postfix(source_table, '_standardization')
+        _create_standardization_table(standardization_table, input_std_table,
+                                      warm_start)
+    else:
+        _create_standardization_table(standardization_table, x_mean_table,
+                                      warm_start)
+        # The original input table is the tab_data_scaled for mini batch.
```
Comment (kaknikhil) [inline src/ports/postgres/modules/convex/mlp_igd.py_in]: this comment should be moved to the if block

File: src/ports/postgres/modules/convex/mlp_igd.py_in [hunk 4]
```diff
@@ -33,11 +34,12 @@ from convex.utils_regularization import __utils_normalize_data_grouping
...
 from utilities.utilities import _array_to_string
+from utilities.utilities import add_postfix
+from utilities.utilities import py_list_to_sql_string as PY2SQL
+from utilities.utilities import extract_keyvalue_params
 from utilities.utilities import _assert
 from utilities.utilities import _assert_equal
 from utilities.utilities import _string_to_array_with_quotes
-from utilities.utilities import add_postfix
-from utilities.utilities import extract_keyvalue_params
 from utilities.utilities import py_list_to_sql_string
```
Comment (kaknikhil) [inline src/ports/postgres/modules/convex/mlp_igd.py_in]: we don't need this import anymore

File: src/ports/postgres/modules/convex/mlp_igd.py_in [hunk 5]
```diff
@@ -292,26 +329,33 @@ def mlp(schema_madlib, source_table, output_table, independent_varname,
...
+        input_std_table = add_postfix(source_table, '_standardization')
+        _create_standardization_table(standardization_table, input_std_table,
+                                      warm_start)
+    else:
+        _create_standardization_table(standardization_table, x_mean_table,
+                                      warm_start)
+        # The original input table is the tab_data_scaled for mini batch.
+        # Do NOT drop this, it will end up dropping the original data table.
+        plpy.execute("DROP TABLE IF EXISTS {0}".format(tbl_data_scaled))
+        plpy.execute("DROP TABLE IF EXISTS {0}".format(x_mean_table))
```
Comment (kaknikhil) [inline src/ports/postgres/modules/convex/mlp_igd.py_in]: is there a test for this in install check to assert that the input tables including summary and std tables aren't dropped for minibatch ?

File: src/ports/postgres/modules/convex/mlp_igd.py_in [hunk 6]
```diff
@@ -491,10 +571,28 @@ def _update_temp_model_table(args, iteration, temp_output_table, first_try):
         ) rel_state_subq
         {join_clause}
     """.format(insert_or_create_str=insert_or_create_str,
-               iteration=iteration, join_clause=join_clause, **args)
+               iteration=iteration, join_clause=join_clause,
+               internal_result_udf=internal_result_udf, **args)
     plpy.execute(model_table_query)
 
 
+def _get_loss(schema_madlib, state, is_mini_batch):
```
Comment (kaknikhil) [inline src/ports/postgres/modules/convex/mlp_igd.py_in]: This function is not used anywhere.

File: src/ports/postgres/modules/convex/mlp_igd.py_in [hunk 7]
```diff
@@ -222,67 +243,83 @@ def mlp(schema_madlib, source_table, output_table, independent_varname,
...
         'group_by_clause': group_by_clause,
         'using_clause': using_clause,
-        'grouping_str_comma': grouping_str_comma
+        'grouping_str_comma': grouping_str_comma,
     })
 
     first_try = True
     temp_output_table = unique_string(desp='temp_output_table')
+
+    layer_sizes = [num_input_nodes] + hidden_layer_sizes + [num_output_nodes]
```
[the remark this answers is not in the document]
Comment (njayaram2) [reply src/ports/postgres/modules/convex/mlp_igd.py_in]: Yes, this looks like duplicated code.

File: src/ports/postgres/modules/convex/mlp_igd.py_in (hunk 1 above)
Comment (njayaram2) [reply src/ports/postgres/modules/convex/mlp_igd.py_in]: Its supposed to be explained in the design doc as per the comment. I think these formulae are taken from a research paper.

File: src/ports/postgres/modules/convex/mlp_igd.py_in (hunk 5 above)
Comment (njayaram2) [reply src/ports/postgres/modules/convex/mlp_igd.py_in]: Yes.

File: src/ports/postgres/modules/convex/mlp_igd.py_in (hunk 6 above)
Comment (njayaram2) [reply src/ports/postgres/modules/convex/mlp_igd.py_in]: Yes, must remove it.

File: src/modules/convex/mlp_igd.cpp [hunk 8]
```diff
@@ -130,6 +145,90 @@ mlp_igd_transition::run(AnyType &args) {
...
+mlp_minibatch_transition::run(AnyType &args) {
+    // For the first tuple: args[0] is nothing more than a marker that
+    // indicates that we should do some initial operations.
+    // For other tuples: args[0] holds the computation state until last tuple
+    MLPMiniBatchState<MutableArrayHandle<double> > state = args[0];
+
+    // initilize the state if first tuple
+    if (state.algo.numRows == 0) {
+        if (!args[3].isNull()) {
+            MLPMiniBatchState<ArrayHandle<double> > previousState = args[3];
```
[the remark this answers is not in the document]
Comment (njayaram2) [reply src/modules/convex/mlp_igd.cpp]: Tried it, it was cleaner this way.

File: src/ports/postgres/modules/convex/test/mlp.sql_in [hunk 9]
```diff
@@ -340,6 +181,51 @@ INSERT INTO iris_data VALUES
...
+\.
+
+-- Create the corresponding summary table for preprocessed data
+CREATE TABLE iris_data_batch_summary(
+    source_table text,
+    output_table text,
+    dependent_varname text,
+    independent_varname text,
+    buffer_size integer,
+    dependent_vartype text,
```
Comment (kaknikhil) [inline src/ports/postgres/modules/convex/test/mlp.sql_in]: Sorry for the late comment but I just realised that we don't need the `dependent_vartype` column in the summary table

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/392/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/395/

**Pull request closed without merging**